File: BJ/BjAlgorizm/BackTracking/15649.cpp
// 22.03.01
// 
// 백트레킹 -> DFS활용 
// 15649
#include <bits/stdc++.h>
using namespace std;


int n, m;
int arr[9];
bool visited[9];

void dfs(int k) {
    if (k==m) {
        for (int i=0; i!=m; i++) {
            cout << arr[i] << " ";
        }
        cout << "\n";
    }
    else {
        for (int i=1; i<=n; i++) {
            if (!visited[i]) { //방문을 하지 않았다면
                visited[i] = true;
                arr[k] = i;
                dfs(k+1);
                visited[i] = false; // 백트래킹 설정
            }
        }
    }
}

int main() {
    cin >> n >> m;
    dfs(0);

}

File: BJ/15000/15649.py
# 순열
n, m = map(int, input().split())
arr = []
visited = [False] * (n + 1)
def dfs(count):
    if count == m:
        for i in range(m):
            print(arr[i], end=' ')
        print()
    else:
        for i in range(1, n+1):
            if not visited[i] :
                visited[i] = True
                arr.append(i)
                dfs(count+1)
                arr.pop()
                visited[i] = False

dfs(0)
# 2. 모듈